### PR TITLE
bpo-32717: Document PEP 560

### PR DESCRIPTION
**Pull request by ilevkivskyi**
I am sorry for a delay! Finally, here are the docs.

Attn: @gvanrossum 

https://bugs.python.org/issue32717

### CODE REVIEW

**Comment by ned-deily** [comment]
Bug fixes appropriate for a maintenance release and doc updates are ok for 370rc1.

**Comment by ilevkivskyi** [comment]
> Bug fixes appropriate for a maintenance release and doc updates are ok for 370rc1.

OK, thanks!

File: Doc/library/types.rst [hunk 1]
```diff
@@ -69,6 +69,23 @@ Dynamic Type Creation
    :pep:`3115` - Metaclasses in Python 3000
       Introduced the ``__prepare__`` namespace hook
 
+.. function:: resolve_bases(bases)
+
+   Resolve MRO entries dynamically as specified by :pep:`560`.
+
+   This function looks for items in ``bases`` that are not instances of
```
Comment (serhiy-storchaka) [inline Doc/library/types.rst]: `*bases*`

Parameters are marked up so.

File: Doc/library/types.rst [hunk 2]
```diff
@@ -69,6 +69,23 @@ Dynamic Type Creation
...
+   :func:`type`, and returns a tuple where each such object that has
```
Comment (serhiy-storchaka) [inline Doc/library/types.rst]: ``` :class:`type` ```?

`:func:` adds `()`.

File: Doc/library/types.rst [hunk 3]
```diff
@@ -69,6 +69,23 @@ Dynamic Type Creation
...
+   an ``__mro_entries__`` method is replaced with an unpacked result of
```
Comment (serhiy-storchaka) [inline Doc/library/types.rst]: I would write ``` :meth:`__mro_entries__` ```.

File: Doc/reference/datamodel.rst [hunk 4]
```diff
@@ -2061,6 +2077,28 @@ case the instance is itself a class.
...
+Emulating generic types
+-----------------------
+
+One can implement the generic class syntax as specified by :pep:`484`
+(for example ``List[int]``) by defining a special method
+
+.. method:: class.__class_getitem__(cls, args)
+
+   Return an object representing the specialization of a generic class
+   by type arguments found in ``args`` tuple.
```
Comment (serhiy-storchaka) [inline Doc/reference/datamodel.rst]: `*args*`

I would rename `args` to `key`. It is not always a tuple.

File: Doc/reference/datamodel.rst [hunk 5]
```diff
@@ -2061,6 +2077,28 @@ case the instance is itself a class.
       module) to the language.
 
 
+Emulating generic types
+-----------------------
+
+One can implement the generic class syntax as specified by :pep:`484`
+(for example ``List[int]``) by defining a special method
+
+.. method:: class.__class_getitem__(cls, args)
```
Comment (serhiy-storchaka) [inline Doc/reference/datamodel.rst]: `object`, not `class`.

I would use `classmethod` as for `__init_subclass__`.

File: Doc/reference/datamodel.rst [hunk 6]
```diff
@@ -2061,6 +2077,28 @@ case the instance is itself a class.
...
+
+   Return an object representing the specialization of a generic class
+   by type arguments found in ``args`` tuple.
+
+This method is looked up on the class object itself, and when defined in
+the class body, this method is implicitly a class method.
+
+.. note:: This mechanism is primarily reserved for use with static type
```
Comment (serhiy-storchaka) [inline Doc/reference/datamodel.rst]: Does it work? This style looks uncommon. I would start the note from the new line.

Or better remove the `note` directive at all. It attracts much attention, but it doesn't look too important.

File: Doc/whatsnew/3.7.rst [hunk 7]
```diff
@@ -356,6 +356,25 @@ with the correct context in async/await code.
        PEP written and implemented by Yury Selivanov
 
 
+PEP 560: Core support for typing module and generic types
+---------------------------------------------------------
+
+Initially :pep:`484` was designed in such way that it would not introduce *any*
+changes to the core CPython interpreter. Now type hints and the :mod:`typing`
+module are extensively used by the community, so this restriction is removed.
+The PEP introduces two special methods ``__class_getitem__`` and
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: ``` :meth:`__class_getitem__` ```

More links is good in What's New.

**Review by serhiy-storchaka — COMMENTED** [review]
Would be nice to add an index entry for `__mro_entries__`.

File: Doc/reference/datamodel.rst (hunk 6 above)
Comment (ilevkivskyi) [reply Doc/reference/datamodel.rst]: It actually worked, but I would just remove it.

File: Doc/library/types.rst (hunk 3 above)
Comment (ilevkivskyi) [reply Doc/library/types.rst]: This actually makes visually inconsistent with how `__prepare__` appears here and elsewhere, so I would keep it as is.

**Comment by ilevkivskyi** [comment]
@serhiy-storchaka Thanks for review! I now addressed your comments, please take a look again.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**